crates/chain/src/tx_graph.rs

```diff
@@ -232,176 +498,446 @@ impl TxGraph {
+                ) => {
+                    additions.txout.extend(
+                        update_txos
+                            .iter()
+                            .filter(|(vout, _)| !txos.contains_key(*vout))
+                            .map(|(&vout, txo)| (OutPoint::new(txid, vout), txo.clone())),
+                    );
+                    *last_seen
                 }
+            };
+
+            if *update_last_seen > prev_last_seen {
+                additions.last_seen.insert(txid, *update_last_seen);
             }
         }
 
+        additions.anchors = update.anchors.difference(&self.anchors).cloned().collect();
+
         additions
     }
+}
 
-    /// Returns the resultant [`Additions`] if the given transaction is inserted. Does not actually
-    /// mutate [`Self`].
+impl<A: Anchor> TxGraph<A> {
+    /// Get all heights that are relevant to the graph.
+    pub fn relevant_heights(&self) -> impl Iterator<Item = u32> + '_ {
+        let mut visited = HashSet::new();
+        self.anchors
+            .iter()
+            .map(|(a, _)| a.anchor_block().height)
+            .filter(move |&h| visited.insert(h))
+    }
+
+    /// Get the position of the transaction in `chain` with tip `chain_tip`.
     ///
-    /// The [`Additions`] result will be empty if `tx` already exists in `self`.
-    pub fn insert_tx_preview(&self, tx: Transaction) -> Additions {
-        let mut update = Self::default();
-        update.txs.insert(tx.txid(), TxNode::Whole(tx));
-        self.determine_additions(&update)
+    /// If the given transaction of `txid` does not exist in the chain of `chain_tip`, `None` is
+    /// returned.
+    ///
+    /// # Error
+    ///
+    /// An error will occur if the [`ChainOracle`] implementation (`chain`) fails. If the
+    /// [`ChainOracle`] is infallible, [`get_chain_position`] can be used instead.
+    ///
+    /// [`get_chain_position`]: Self::get_chain_position
+    pub fn try_get_chain_position<C>(
+        &self,
+        chain: &C,
+        chain_tip: BlockId,
+        txid: Txid,
+    ) -> Result<Option<ObservedAs<&A>>, C::Error>
+    where
+        C: ChainOracle,
+    {
+        let (tx_node, anchors, &last_seen) = match self.txs.get(&txid) {
+            Some((tx, anchors, last_seen)) if !(anchors.is_empty() && *last_seen == 0) => {
+                (tx, anchors, last_seen)
+            }
+            _ => return Ok(None),
+        };
+
+        for anchor in anchors {
+            match chain.is_block_in_chain(anchor.anchor_block(), chain_tip)? {
+                Some(true) => return Ok(Some(ObservedAs::Confirmed(anchor))),
+                _ => continue,
+            }
+        }
+
+        // If we cannot determine whether tx is in best chain, we can check whether a spending tx is
+        // confirmed and in best chain, and if so, it is guaranteed that this tx is in the best
+        // chain.
+        //
+        // [TODO] This logic is incomplete as we do not check spends of spends.
```

Comment:
Can we remove this logic and the whole `ConfirmedImplicit`. I don't think we want half-completed things.